src/typechecker/FStar.TypeChecker.TcTerm.fs

```diff
@@ -1302,9 +1339,18 @@ and tc_match (env : Env.env) (top : term) : term * lcomp * guard_t =
     in
 
     let e =
+      //repack the returns ascription
+      //we make the binder sort as tun,
+      //  since we always use the type of the scrutinee
+      let ret_opt =
+        match ret_opt with
+        | None -> None
+        | Some (b, asc) ->
+          let asc = SS.close_ascription [b] asc in
+          let b = List.hd (SS.close_binders [b]) in
+          let b = {b with binder_bv={b.binder_bv with sort=tun}} in
```

Comment:
While packing the returns annotation back, set the sort to `_`. So we don't have to normalize it, instantiate it, compute its fvs etc. etc.

Reply:
interesting ... nice trick